quam/components/ports.py

```diff
@@ -0,0 +1,266 @@
+                        f"Overwriting {port_config['key']} → {value}."
+                    )
+            except Exception:
+                pass
+            port_config[key] = value
+
+    def apply_to_config(self, config: Dict) -> None:
+        super().apply_to_config(config)
+
+        port_cfg = self.get_port_config(config)
+        port_properties = self.get_port_properties()
+        self._update_port_config(port_cfg, port_properties)
+
+
+@quam_dataclass
+class OPXPlusPort(Port, ABC):
+    port: Tuple[str, int]
+
+    def get_port_config(
+        self, config: Dict[str, Any], create: bool = True
+    ) -> Dict[str, Any]:
+        controller_name, port = self.port
+
+        if not create:
+            try:
+                return config["controllers"][controller_name][f"{self.port_type}"][port]
+            except KeyError:
+                raise KeyError(
+                    f"Error generating config: controller {controller_name} does not "
+                    f"have entry {self.port_type}s for port {self.port}"
+                )
+
+        controller_cfg = config["controllers"].setdefault(controller_name, {})
+        ports_cfg = controller_cfg.setdefault(f"{self.port_type}s", {})
+        port_cfg = ports_cfg.setdefault(port, {})
+        return port_cfg
+
+
+@quam_dataclass
+class FEMPort(Port, ABC):
+    port: Tuple[str, int, int]
+
+    def get_port_config(
+        self, config: Dict[str, Any], create: bool = True
+    ) -> Dict[str, Any]:
+        controller_name, fem, port = self.port
+
+        if not create:
+            try:
+                fem_cfg = config["controllers"][controller_name]["fems"][fem]
+            except KeyError:
+                raise KeyError(
+                    f"Error generating config: controller {controller_name} does not "
+                    f"have entry for FEM {fem} for port {self.port}"
+                )
+            try:
+                return fem_cfg[f"{self.port_type}s"][port]
+            except KeyError:
+                raise KeyError(
+                    f"Error generating config: controller {controller_name} does not "
+                    f"have entry {self.port_type}s for port {self.port}"
+                )
+
+        controller_cfg = config["controllers"].setdefault(controller_name, {})
+        fems_cfg = controller_cfg.setdefault("fems", {})
+        fem_cfg = fems_cfg.setdefault(fem, {})
+        ports_cfg = fem_cfg.setdefault(f"{self.port_type}s", {})
+        port_cfg = ports_cfg.setdefault(port, {})
+        return port_cfg
+
+
+# --- Analog ports --- #
+@quam_dataclass
+class LFAnalogOutputPort(QuamComponent, ABC):
+    port_type: ClassVar[str] = "analog_output"
+
+    offset: float = 0.0
+    delay: int = 0
+    crosstalk: Dict[int, float] = field(default_factory=dict)
+    feedforward_filter: List[float] = field(default_factory=list)
+    feedback_filter: List[float] = field(default_factory=list)
```

Comment:
I am not sure what would be the behavior here.
Would take put empty lists as the default values?
If so, this **might** create the wrong behavior and add significant delay to all channels (as if there are filters & feedback).
The default value should be "None"

Reply:
Okay so an empty list/dict would be treated differently than if the entry didn't exist? I wasn't aware of this. Should we also include this information in the qua config document on Confluence?

Reply:
I'm not 100% sure, but I would suspect it might be the case.

On confluence It says the default is None, I am not sure what happens if an empty list is added...
I'm open to ideas on how to write it better.

Reply:
Set default values to None